// @jiaminghi/transition/config.js
export const linear = [[[0, 1], '', [0.33, 0.67]], [[1, 0], [0.67, 0.33]]]

export const easeInSine = [[[0, 1]],[[0.5386100386100386, 0.564], [0.16988416988416988, 0.912], [0.8803088803088803, 0.196]],[[1, 0]]]

export const easeOutSine = [[[0, 1]],[[0.4691119691119691, 0.456], [0.12355212355212356, 0.796], [0.8146718146718147, 0.116]],[[1, 0]]]

export const easeInOutSine = [[[0, 1]],[[0.49613899613899615, 0.5], [0.24517374517374518, 0.988], [0.747104247104247, 0.012]],[[1, 0]]]




export const easeInQuad = [[[0, 1]],[[0.5501930501930502, 0.584], [0.23166023166023167, 0.904], [0.8687258687258688, 0.264]],[[1, 0]]]

export const easeOutQuad = [[[0, 1]],[[0.41312741312741313, 0.428], [0.06563706563706563, 0.816], [0.7606177606177607, 0.04]],[[1, 0]]]

export const easeInOutQuad = [[[0, 1]],[[0.49613899613899615, 0.488], [0.32432432432432434, 0.968], [0.6235521235521235, 0.056]],[[1, 0]]]




export const easeInCubic = [[[0, 1]],[[0.6795366795366795, 0.688], [0.3667953667953668, 0.992], [0.9922779922779923, 0.384]],[[1, 0]]]

export const easeOutCubic = [[[0, 1]],[[0.333976833976834, 0.216], [0.03861003861003861, 0.536], [0.6235521235521235, -0.024]],[[1, 0]]]

export const easeInOutCubic = [[[0, 1]],[[0.4980694980694981, 0.504], [0.2722007722007722, 1.036], [0.6737451737451737, -0.028]],[[1, 0]]]




export const easeInQuart = [[[0, 1]],[[0.803088803088803, 0.772], [0.6081081081081081, 1.052], [0.9980694980694981, 0.492]],[[1, 0]]]

export const easeOutQuart = [[[0, 1]],[[0.1525096525096525, 0.244], [0.0019305019305019305, 0.448], [0.2857142857142857, -0.02]],[[1, 0]]]

export const easeInOutQuart = [[[0, 1]],[[0.4942084942084942, 0.508], [0.4343629343629344, 1.056], [0.5463320463320464, -0.064]],[[1, 0]]]




export const easeInQuint = [[[0, 1]],[[0.8996138996138996, 0.872], [0.7895752895752896, 0.996], [1.0096525096525097, 0.748]],[[1, 0]]]

export const easeOutQuint = [[[0, 1]],[[0.09073359073359073, 0.168], [-0.003861003861003861, 0.388], [0.22586872586872586, -0.004]],[[1, 0]]]

export const easeInOutQuint = [[[0, 1]],[[0.4826254826254826, 0.504], [0.5366795366795367, 1.06], [0.44015444015444016, -0.052]],[[1, 0]]]




export const easeInExpo = [[[0, 1]],[[0.8976833976833977, 0.864], [0.8204633204633205, 1.008], [1.0135135135135136, 0.764]],[[1, 0]]]

export const easeOutExpo = [[[0, 1]],[[0.0694980694980695, 0.128], [0.005791505791505791, 0.216], [0.10617760617760617, -0.012]],[[1, 0]]]

export const easeInOutExpo = [[[0, 1]],[[0.48455598455598453, 0.512], [0.5231660231660231, 1.004], [0.44594594594594594, 0.016]],[[1, 0]]]




export const easeInBack = [[[0, 1]],[[0.667953667953668, 0.896], [0.3803088803088803, 1.184], [0.9555984555984556, 0.616]],[[1, 0]]]

export const easeOutBack = [[[0, 1]],[[0.3359073359073359, 0.028], [0.06177606177606178, 0.22], [0.6312741312741312, -0.18]],[[1, 0]]]

export const easeInOutBack = [[[0, 1]],[[0.48841698841698844, 0.484], [0.36486486486486486, 1.404], [0.611969111969112, -0.436]],[[1, 0]]]



export const easeInElastic = [[[0, 1]],[[0.48841698841698844, 0.944], [0.4015444015444015, 1.124], [0.5752895752895753, 0.764]],[[0.6891891891891891, 1.124], [0.5540540540540541, 0.996], [0.832046332046332, 1.048]],[[1, 0]]]

export const easeOutElastic = [[[0, 1]],[[0.16216216216216217, -0.128], [-0.032818532818532815, 0.164], [0.28378378378378377, -0.136]],[[0.34555984555984554, 0.104], [0.21621621621621623, 0.124], [0.5386100386100386, 0]],[[1, 0]]]

export const easeInOutElastic = [[[0, 1]],[[0.07528957528957529, 1.06], [0.0444015444015444, 1.088], [0.16602316602316602, 1.04]],[[0.2528957528957529, 1.116], [0.15444015444015444, 0.724], [0.3745173745173745, 1.116]],[[0.666023166023166, -0.184], [0.4826254826254826, 0.052], [0.7432432432432432, -0.14]],[[0.8011583011583011, 0.088], [0.693050193050193, 0.052], [0.9247104247104247, -0.156]],[[1, 0]]]

export default new Map([
  ['linear', linear],
  ['easeInSine', easeInSine],
  ['easeOutSine', easeOutSine],
  ['easeInOutSine', easeInOutSine],
  ['easeInQuad', easeInQuad],
  ['easeOutQuad', easeOutQuad],
  ['easeInOutQuad', easeInOutQuad],
  ['easeInCubic', easeInCubic],
  ['easeOutCubic', easeOutCubic],
  ['easeInOutCubic', easeInOutCubic],
  ['easeInQuart', easeInQuart],
  ['easeOutQuart', easeOutQuart],
  ['easeInOutQuart', easeInOutQuart],
  ['easeInQuint', easeInQuint],
  ['easeOutQuint', easeOutQuint],
  ['easeInOutQuint', easeInOutQuint],
  ['easeInExpo', easeInExpo],
  ['easeOutExpo', easeOutExpo],
  ['easeInOutExpo', easeInOutExpo],
  ['easeInBack', easeInBack],
  ['easeOutBack', easeOutBack],
  ['easeInOutBack', easeInOutBack],
  ['easeInElastic', easeInElastic],
  ['easeOutElastic', easeOutElastic],
  ['easeInOutElastic', easeInOutElastic]
])
